clang/test/CIR/Transforms/mem2reg.cir

```diff
@@ -0,0 +1,206 @@
+      cir.if %12 {
+        %13 = cir.alloca !s32i, !cir.ptr<!s32i>, ["z", init] {alignment = 4 : i64}
+        %14 = cir.const #cir.int<2> : !s32i
+        cir.store %14, %13 : !s32i, !cir.ptr<!s32i>
+        %15 = cir.load %0 : !cir.ptr<!s32i>, !s32i
+        %16 = cir.load %13 : !cir.ptr<!s32i>, !s32i
+        %17 = cir.binop(mul, %15, %16) : !s32i
+        cir.store %17, %2 : !s32i, !cir.ptr<!s32i>
+      } else {
+        %13 = cir.alloca !s32i, !cir.ptr<!s32i>, ["z", init] {alignment = 4 : i64}
+        %14 = cir.const #cir.int<3> : !s32i
+        cir.store %14, %13 : !s32i, !cir.ptr<!s32i>
+        %15 = cir.load %0 : !cir.ptr<!s32i>, !s32i
+        %16 = cir.load %13 : !cir.ptr<!s32i>, !s32i
+        %17 = cir.binop(mul, %15, %16) : !s32i
+        cir.store %17, %2 : !s32i, !cir.ptr<!s32i>
+     }
+    }
+    %4 = cir.load %2 : !cir.ptr<!s32i>, !s32i
+    %5 = cir.const #cir.int<1> : !s32i
+    %6 = cir.binop(add, %4, %5) : !s32i
+    cir.store %6, %2 : !s32i, !cir.ptr<!s32i>
+    %7 = cir.load %2 : !cir.ptr<!s32i>, !s32i
+    cir.store %7, %1 : !s32i, !cir.ptr<!s32i>
+    %8 = cir.load %1 : !cir.ptr<!s32i>, !s32i
+    cir.return %8 : !s32i
+  }
+  // CHECK:  cir.func @alloca_in_ifelse(%arg0: !s32i) -> !s32i {
+  // CHECK:    %0 = cir.const #cir.int<0> : !s32i
+  // CHECK:    cir.br ^bb1
+  // CHECK:  ^bb1:  // pred: ^bb0
+  // CHECK:    %1 = cir.const #cir.int<42> : !s32i
+  // CHECK:    %2 = cir.cmp(gt, %arg0, %1) : !s32i, !s32i
+  // CHECK:    %3 = cir.cast(int_to_bool, %2 : !s32i), !cir.bool
+  // CHECK:    cir.brcond %3 ^bb3, ^bb2
+  // CHECK:  ^bb2:  // pred: ^bb1
+  // CHECK:    %4 = cir.const #cir.int<3> : !s32i
+  // CHECK:    %5 = cir.binop(mul, %arg0, %4) : !s32i
+  // CHECK:    cir.br ^bb4(%5 : !s32i)
+  // CHECK:  ^bb3:  // pred: ^bb1
+  // CHECK:    %6 = cir.const #cir.int<2> : !s32i
+  // CHECK:    %7 = cir.binop(mul, %arg0, %6) : !s32i
+  // CHECK:    cir.br ^bb4(%7 : !s32i)
+  // CHECK:  ^bb4(%8: !s32i):  // 2 preds: ^bb2, ^bb3
+  // CHECK:    cir.br ^bb5
+  // CHECK:  ^bb5:  // pred: ^bb4
+  // CHECK:    %9 = cir.const #cir.int<1> : !s32i
+  // CHECK:    %10 = cir.binop(add, %8, %9) : !s32i
+  // CHECK:    cir.return %10 : !s32i
+  // CHECK:  }
+
+  //  ====   Alloca promoted through bitcast
+  // C code
+  // typedef __SIZE_TYPE__ size_t;
+  // void *alloca(size_t size);
+  // 
+  // void test_bitcast(size_t n) {
+  //   int *c1 = alloca(n);
+  // }
+  cir.func @test_bitcast(%arg0: !u64i) {
+    %0 = cir.alloca !u64i, !cir.ptr<!u64i>, ["n", init] {alignment = 8 : i64}
+    %1 = cir.alloca !cir.ptr<!s32i>, !cir.ptr<!cir.ptr<!s32i>>, ["c1", init] {alignment = 8 : i64}
+    cir.store %arg0, %0 : !u64i, !cir.ptr<!u64i>
+    %2 = cir.load %0 : !cir.ptr<!u64i>, !u64i
+    %3 = cir.alloca !u8i, !cir.ptr<!u8i>, %2 : !u64i, ["bi_alloca"] {alignment = 16 : i64}
+    %4 = cir.cast(bitcast, %3 : !cir.ptr<!u8i>), !cir.ptr<!void>
+    %5 = cir.cast(bitcast, %4 : !cir.ptr<!void>), !cir.ptr<!s32i>
```

Comment:
I see now how the bitcast is leveraged, awesome. Still curious about that question though.

Reply:
well, as `CastOp` doesn't touch memory indeed! unless it's a `bitcast`. It doesn't read or write memory though, but still may propagate a pointer somehow. So the bitcast itself is transparent in that sense that the decision if such kind of operation can be removed depends on its users only but not from the cast operation.